system-helper/flatpak-system-helper.c

```diff
@@ -1151,6 +1289,215 @@ handle_run_triggers (FlatpakSystemHelper   *object,
+  g_autoptr(GVariant) unixUserId = NULL;
+  g_autoptr(GVariant) credentials = NULL;
+  uid_t uid;
+
+   credentials = g_dbus_connection_call_sync (connection,
+                                             "org.freedesktop.DBus",
+                                             "/org/freedesktop/DBus",
+                                             "org.freedesktop.DBus",
+                                             "GetConnectionCredentials",
+                                             g_variant_new ("(s)", sender),
+                                             G_VARIANT_TYPE ("(a{sv})"), G_DBUS_CALL_FLAGS_NONE, G_MAXINT, NULL, error);
+  if (credentials == NULL)
+    return FALSE;
+
+   dict = g_variant_get_child_value (credentials, 0);
+  unixUserId = g_variant_lookup_value (dict, "UnixUserID", G_VARIANT_TYPE_UINT32);
+
+   if (unixUserId == NULL)
+    {
+      g_set_error (error, G_DBUS_ERROR, G_DBUS_ERROR_FAILED,
+                   "Failed to query UnixUserId for the bus name: %s", sender);
+      return FALSE;
+    }
+
+   *uid_out = g_variant_get_uint32 (unixUserId);
+  return TRUE;
+}
+
+
+static void
+revokefs_child_setup (gpointer user_data)
+{
+  struct passwd *passwd = getpwnam("flatpak");
+  setgid (passwd->pw_gid);
+  setuid (passwd->pw_uid);
+
+  close (GPOINTER_TO_INT(user_data));
+}
+
+static void
+name_vanished_cb (GDBusConnection *connection, const gchar *name, gpointer user_data)
+{
+  OngoingPull *pull = (OngoingPull *) user_data;
+
+  ongoing_pull_free (pull);
+}
+
+static OngoingPull *
+ongoing_pull_new (FlatpakSystemHelper   *object,
+                  GDBusMethodInvocation *invocation,
+                  const gchar           *ref,
+                  const gchar           *commit,
+                  uid_t                  uid,
+                  gchar                 *revokefs_fuse_path,
+                  gchar                 *src,
+                  gchar                 *mnt,
+                  GError               **error)
+{
+  GDBusConnection *connection = g_dbus_method_invocation_get_connection (invocation);
+  OngoingPull *pull;
+  g_autoptr (GSubprocessLauncher) launcher = NULL;
+  g_autoptr (GSubprocess) revokefs_backend = NULL;
+  g_autofree gchar *target_socket = NULL;
+  int sockets[2];
+
+  pull = g_slice_new0 (OngoingPull);
+  pull->object = object;
+  pull->invocation = invocation;
+  pull->ref = g_strdup (ref);
+  pull->commit = g_strdup (commit);
+  pull->src_dir = g_strdup (src);
+  pull->mountpoint = g_strdup (mnt);
+  pull->cancellable = g_cancellable_new ();
+  pull->uid = uid;
+
+  pull->watch_id = g_bus_watch_name_on_connection (connection,
+                                                   g_dbus_connection_get_unique_name (connection),
+                                                   G_BUS_NAME_WATCHER_FLAGS_NONE, NULL,
+                                                   name_vanished_cb,
+                                                   pull, NULL); /* TODO: Take a ref here */
```

Comment:
TODO

Reply:
Is a ref actually needed, if we remove `watch_id` when the OngoingPull struct is destroyed? I would hope not.

Reply:
There actually is an issue here. Consider the case where a client calls `Deploy` and then immediately crashes. Then, in the system helper, there are two threads:

1. the thread handling Deploy, which will remove `pull` from the list (ostensibly taking ownership of it), do some stuff which may take a long time, then free it
2. the main thread, which will get `name_vanished_cb` called and try to free `pull`

Boom – double-free.

It would be nice to avoid ref-counting – perhaps it's enough for `name_vanished_cb` to claim the `ongoing_pulls` lock, check whether its `pull` pointer is still in the array and if so remove it (taking ownership), unlock, and then free it if it was found.

Reply:
Against my own point there: the memory pointed at by `pull` could be re-used by a future pull. Another option is to store the unique name of the connection in the `pull` structure, pass `g_strdup (g_dbus_connection_get_unique_name (connection)), g_free` as the userdata and free func to watch_name__on_connection, and then in the callback, claim the ongoing_pulls lock, iterate through it destroying any ongoing pull for that unique name, and then release the lock.